plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/AtomicFUTransformer.kt

```diff
@@ -0,0 +1,595 @@
+        "AtomicBooleanArray" to "BooleanArray",
+        "AtomicArray" to "Array"
+    )
+
+    override fun visitFile(declaration: IrFile): IrFile {
+        val transformedDeclarations = mutableListOf<IrDeclaration>()
+        declaration.declarations.forEach { irDeclaration ->
+            irDeclaration.transformInlineAtomicExtension()?.let { transformedDeclarations.add(it) }
+        }
+        declaration.declarations.addAll(transformedDeclarations)
+        return super.visitFile(declaration)
+    }
+
+    override fun visitClass(declaration: IrClass): IrStatement {
+        val transformedDeclarations = mutableListOf<IrDeclaration>()
+        declaration.declarations.forEach { irDeclaration ->
+            irDeclaration.transformInlineAtomicExtension()?.let { transformedDeclarations.add(it) }
+        }
+        declaration.declarations.addAll(transformedDeclarations)
+        return super.visitClass(declaration)
+    }
+
+    override fun visitProperty(declaration: IrProperty): IrStatement {
+        if (declaration.backingField != null) {
+            val backingField = declaration.backingField!!
+            if (backingField.initializer != null) {
+                val initializer = backingField.initializer!!.expression.eraseAtomicConstructor(backingField)
+                declaration.backingField!!.initializer = context.irFactory.createExpressionBody(initializer)
+            }
+        }
+        return super.visitProperty(declaration)
+    }
+
+    override fun visitFunction(declaration: IrFunction): IrStatement {
+        transformDeclarationBody(declaration.body, declaration)
+        return super.visitFunction(declaration)
+    }
+
+    override fun visitAnonymousInitializer(declaration: IrAnonymousInitializer): IrStatement {
+        transformDeclarationBody(declaration.body, declaration)
+        return super.visitAnonymousInitializer(declaration)
+    }
+
+    private fun transformDeclarationBody(body: IrBody?, parent: IrDeclaration, lambda: IrFunction? = null) {
+        if (body is IrBlockBody) {
+            body.statements.forEachIndexed { i, stmt ->
+                val transformedStmt = stmt.transformStatement(parent, lambda)
+                body.statements[i] = transformedStmt
+            }
+        }
+    }
+
+    private fun IrExpression.eraseAtomicConstructor(parentDeclaration: IrDeclaration, lambda: IrFunction? = null) =
+        when {
+            type.isAtomicValueType() -> getPureTypeValue().transformAtomicFunctionCall(parentDeclaration, lambda)
+            type.isAtomicArrayType() -> buildPureTypeArrayConstructor()
+            type.isReentrantLockType() -> buildConstNull()
+            else -> this
+        }
+
+    private fun IrDeclaration.transformInlineAtomicExtension(): IrDeclaration? {
+        // transform the signature of the inline Atomic* extension declaration
+        // inline fun AtomicRef<T>.foo(args) { ... } -> inline fun foo(args, getter: () -> T, setter: (T) -> Unit)
```

Comment:
It has to be
`inline fun <T> AtomicRef<T>.foo(args) { ... } -> inline fun <T> foo(args, getter: () -> T, setter: (T) -> Unit)`